File: wpilibc/src/test/native/cpp/SolenoidTestREV.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>

#include "wpi/hardware/pneumatic/DoubleSolenoid.hpp"
#include "wpi/hardware/pneumatic/Solenoid.hpp"

namespace wpi {
TEST_CASE("SolenoidREVTest ValidInitialization", "[wpilibc]") {
  Solenoid solenoid{CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 2};
  CHECK(2 == solenoid.GetChannel());

  solenoid.Set(true);
  CHECK(solenoid.Get());

  solenoid.Set(false);
  CHECK_FALSE(solenoid.Get());
}

TEST_CASE("SolenoidREVTest DoubleInitialization", "[wpilibc]") {
  Solenoid solenoid{CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 2};
  CHECK_THROWS_AS(
      Solenoid(CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 2),
      std::runtime_error);
}

TEST_CASE("SolenoidREVTest DoubleInitializationFromDoubleSolenoid",
          "[wpilibc]") {
  DoubleSolenoid solenoid{CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH,
                          2, 3};
  CHECK_THROWS_AS(
      Solenoid(CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 2),
      std::runtime_error);
}

TEST_CASE("SolenoidREVTest InvalidChannel", "[wpilibc]") {
  CHECK_THROWS_AS(
      Solenoid(CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 100),
      std::runtime_error);
}

TEST_CASE("SolenoidREVTest Toggle", "[wpilibc]") {
  Solenoid solenoid{CANPort::CAN_S0, 3, wpi::PneumaticsModuleType::REV_PH, 2};
  solenoid.Set(true);
  CHECK(solenoid.Get());

  solenoid.Toggle();
  CHECK_FALSE(solenoid.Get());

  solenoid.Toggle();
  CHECK(solenoid.Get());
}
}  // namespace wpi

File: wpilibj/src/test/java/org/wpilib/hardware/pneumatic/SolenoidTestREV.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.hardware.pneumatic;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;
import org.wpilib.hardware.hal.util.AllocationException;

class SolenoidTestREV {
  @Test
  void testValidInitialization() {
    try (Solenoid solenoid = new Solenoid(0, 3, PneumaticsModuleType.REVPH, 2)) {
      assertEquals(2, solenoid.getChannel());

      solenoid.set(true);
      assertTrue(solenoid.get());

      solenoid.set(false);
      assertFalse(solenoid.get());
    }
  }

  @Test
  void testDoubleInitialization() {
    try (Solenoid solenoid = new Solenoid(0, 3, PneumaticsModuleType.REVPH, 2)) {
      assertThrows(
          AllocationException.class, () -> new Solenoid(0, 3, PneumaticsModuleType.REVPH, 2));
    }
  }

  @Test
  void testDoubleInitializationFromDoubleSolenoid() {
    try (DoubleSolenoid solenoid = new DoubleSolenoid(0, 3, PneumaticsModuleType.REVPH, 2, 3)) {
      assertThrows(
          AllocationException.class, () -> new Solenoid(0, 3, PneumaticsModuleType.REVPH, 2));
    }
  }

  @Test
  void testInvalidChannel() {
    assertThrows(
        IllegalArgumentException.class, () -> new Solenoid(0, 3, PneumaticsModuleType.REVPH, 100));
  }

  @Test
  void testToggle() {
    try (Solenoid solenoid = new Solenoid(0, 3, PneumaticsModuleType.REVPH, 2)) {
      solenoid.set(true);
      assertTrue(solenoid.get());

      solenoid.toggle();
      assertFalse(solenoid.get());

      solenoid.toggle();
      assertTrue(solenoid.get());
    }
  }
}
